### PR TITLE
Fix memory leak in engine_cleanup_add_first

### LINKED ISSUE #20870: memory leak in engine_cleanup_add_first
The engine_cleanup_add_first function in the crypto/engine/eng_lib.c source file appears to contain a memory leak. Specifically, memory is allocated using int_cleanup_item in line 142:

item = int_cleanup_item(cb);

However, if the sk_ENGINE_CLEANUP_ITEM_insert function fails to add the item object to the stack, the application should take appropriate action to clean up the allocated memory to avoid memory leaks.

### PR DESCRIPTION
**Pull request by KovalevVasiliy**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by KovalevVasiliy** [comment]
@tom-cosgrove-arm Do I really need to sign a CLA and send it back tolegal@opensslfoundation.org?

File: crypto/engine/eng_lib.c [hunk 1]
```diff
@@ -140,8 +140,11 @@ void engine_cleanup_add_first(ENGINE_CLEANUP_CB *cb)
     if (!int_cleanup_check(1))
         return;
     item = int_cleanup_item(cb);
-    if (item)
-        sk_ENGINE_CLEANUP_ITEM_insert(cleanup_stack, item, 0);
+    if (item != NULL) {
+        if (sk_ENGINE_CLEANUP_ITEM_insert(cleanup_stack, item, 0) <= 0) {
+            OPENSSL_free(item);
+        }
+    }
```
Comment (t8m) [inline crypto/engine/eng_lib.c:147]: Style nit, we do not add curly brackets when there is just a single statement inside. Please remove the curly brackets.

Comment (KovalevVasiliy) [reply crypto/engine/eng_lib.c]: > Style nit, we do not add curly brackets when there is just a single statement inside. Please remove the curly brackets.

Fixed

File: crypto/engine/eng_lib.c [hunk 2]
```diff
@@ -140,8 +140,10 @@ void engine_cleanup_add_first(ENGINE_CLEANUP_CB *cb)
     if (!int_cleanup_check(1))
         return;
     item = int_cleanup_item(cb);
-    if (item)
-        sk_ENGINE_CLEANUP_ITEM_insert(cleanup_stack, item, 0);
+    if (item != NULL) {
```
Comment (t8m) [inline crypto/engine/eng_lib.c:143]: Technically these curly brackets aren't needed either but we are more liberal about such cases.

Comment (KovalevVasiliy) [reply crypto/engine/eng_lib.c:143]: I did it similar to engine_cleanup_add_last.

Comment (KovalevVasiliy) [reply crypto/engine/eng_lib.c:143]: > Technically these curly brackets aren't needed either but we are more liberal about such cases.

Fixed

**Review by t8m — APPROVED** [review]

**Comment by tom-cosgrove-arm** [comment]
Setting `approval: done` - I presume this is separate from handling of the CLA

**Review by mattcaswell — APPROVED** [review]

**Comment by KovalevVasiliy** [comment]
I've got an email, what I has been added to CLA. I've got an instruction, what I need to close and reopen this PR, but it didn't help

**Comment by tom-cosgrove-arm** [comment]
Does the email address in the ICLA match that in the commit (which is `minskij10@yandex.ru`)?

**Comment by t8m** [comment]
Merged to master, 3.1, and 3.0 branches after improving the commit message. Thank you for your contribution.

**Pull request closed without merging**